lua/SS13_base.lua

```diff
@@ -67,123 +46,92 @@ function SS13.await(thing_to_call, proc_to_call, ...)
+		"__SS13_signal_handlers",
+		dm.global_procs.WEAKREF(datum),
+		"parent_qdeleting",
+		"handler",
+	}
+	callback.arguments = { path }
+	local handler_data = { callback = callback, functions = {} }
+	handler_data.handler = function(source, ...)
+		local result = signal_handler(handler_data, source, ...)
+		for signal, signal_data in __SS13_signal_handlers[source] do
+			signal_data.callback:UnregisterSignal(source, signal)
 		end
-		SS13.stop_tracking(handler_callback)
-	end
-
-	local datumWeakRef = datum
-	if not SS13.istype(datum, "/datum/weakref") then
-		datumWeakRef = dm.global_proc("WEAKREF", datum)
-	end
-	if not __SS13_signal_handlers[datumWeakRef] then
-		return
+		__SS13_signal_handlers[source] = nil
+		return result
 	end
+	__SS13_signal_handlers[datum]["parent_qdeleting"] = handler_data
+end
 
-	if signal == "_cleanup" then
+function SS13.register_signal(datum, signal, func)
+	if not SS13.istype(datum, "/datum") then
 		return
 	end
-
-	if not __SS13_signal_handlers[datumWeakRef][signal] then
-		return
+	if not SS13.is_valid(datum) then
+		error("Tried to register a signal on a deleted datum", 2)
 	end
-
-	if not callback then
-		for handler_key, handler_info in __SS13_signal_handlers[datumWeakRef][signal] do
-			clear_handler(handler_info)
+	if not __SS13_signal_handlers[datum] then
+		__SS13_signal_handlers[datum] = {}
+		-- Turfs don't remove their signals on deletion.
+		if not SS13.istype(datum, "/turf") then
+			create_qdeleting_callback(datum)
 		end
-		__SS13_signal_handlers[datumWeakRef][signal] = nil
-	else
-		if not SS13.istype(callback, "/datum/callback") then
-			return
+	end
+	local handler_data = __SS13_signal_handlers[datum][signal]
+	if not handler_data then
+		handler_data = { callback = nil, functions = {} }
+		local callback = SS13.new("/datum/callback", SS13.state, "call_function_return_first")
+		callback:RegisterSignal(datum, signal, "Invoke")
+		local path = {
+			"__SS13_signal_handlers",
+			dm.global_procs.WEAKREF(datum),
+			signal,
+			"handler",
+		}
+		callback.arguments = { path }
+		handler_data.callback = callback
+		handler_data.handler = function(...)
+			return signal_handler(handler_data, ...)
 		end
-		local callbackWeakRef = dm.global_proc("WEAKREF", callback)
-		clear_handler(__SS13_signal_handlers[datumWeakRef][signal][callbackWeakRef])
-		__SS13_signal_handlers[datumWeakRef][signal][callbackWeakRef] = nil
+		__SS13_signal_handlers[datum][signal] = handler_data
 	end
+	handler_data.functions[func] = true
+	return true
 end
 
-function SS13.signal_handler_cleanup(datumWeakRef)
-	if not __SS13_signal_handlers[datumWeakRef] then
+function SS13.unregister_signal(datum, signal, func)
+	if not __SS13_signal_handlers[datum] then
 		return
 	end
-
-	for signal, _ in __SS13_signal_handlers[datumWeakRef] do
-		SS13.unregister_signal(datumWeakRef, signal)
+	local handler_data = __SS13_signal_handlers[datum][signal]
+	if not handler_data then
+		return
+	end
+	handler_data.functions[func] = nil
```

Comment:
This causes a runtime if func is null.
The default behaviour (as set out before) should be that all signals on datum for signaltype `signal` should be unregistered. Otherwise, this is going to break the majority of my lua scripts